lumicks/pylake/recording.py

```diff
@@ -0,0 +1,154 @@
+import numpy as np
+import os
+import re
+
+from .channel import Slice, Continuous, TimeSeries
+
+class TiffFrame:
+    """Thin wrapper around a TIFF frame stack. For camera videos timestamps are stored in the DateTime tag in
+    the format start:end.
+
+    Parameters
+    ----------
+    page : tifffile.tifffile.TiffPage
+        Tiff page recorded from a camera in Bluelake.
+    """
+    def __init__(self, page):
+        self._src = page
+
+    @property
+    def data(self):
+        return self._src.asarray()
+
+    @property
+    def start(self):
+        timestamp_string = re.search('^(\d+):\d+$', self._src.tags['DateTime'].value)
+        return np.int64(timestamp_string.group(1)) if timestamp_string else None
+
+    @property
+    def stop(self):
+        timestamp_string = re.search('^\d+:(\d+)$', self._src.tags['DateTime'].value)
+        return np.int64(timestamp_string.group(1)) if timestamp_string else None
+
+
+class TiffStack:
+    """TIFF images exported from Bluelake
+
+        Parameters
+        ----------
+        tiff_file : tifffile.TiffFile
+            Filename pointing to a TIFF file recorded from a camera in Bluelake.
+        """
+    def __init__(self, tiff_file):
+        self._src = tiff_file
+
+    def get_frame(self, frame):
+        return TiffFrame(self._src.pages[frame])
+
+    @staticmethod
+    def from_file(image_file):
+        import tifffile
+        return TiffStack(tifffile.TiffFile(image_file))
+
+    @property
+    def num_frames(self):
+        return len(self._src.pages)
+
+
+class Recording:
+    """Recording obtained with Bluelake
+
+        Parameters
+        ----------
+        data : TiffStack
+            TiffStack object.
+        name : str
+            Plot label of the recording
+        start_idx : int
+            Index at the first frame.
+        stop_idx: int
+            Index beyond the last frame.
+        """
+    def __init__(self, data, name=None, start_idx=0, stop_idx=None):
+        self.src = data
+        self.name = name
+        self.start_idx = start_idx
+        self.stop_idx = (self.src.num_frames if stop_idx is None else stop_idx)
+
+    def __getitem__(self, item):
+        """All indexing is in frames"""
+        if isinstance(item, slice):
+            if item.step is not None:
+                raise IndexError("Slice steps are not supported")
+
+            start, stop, _ = item.indices(self.num_frames)
+            return Recording(self.src, self.name, self.start_idx + start, self.start_idx + stop)
+        else:
+            item = self.start_idx + item if item >= 0 else self.stop_idx + item
+            if item >= self.stop_idx or item < self.start_idx:
+                raise IndexError("Index out of bounds")
+            return Recording(self.src, self.name, item, item + 1)
+
+    @staticmethod
+    def from_file(image_name):
```

Comment:
I do agree that syntax is nicer and it crossed my mind as well. I was actually considering making a mock function Recording which calls from_file and returns the object, but I figured that'd also look pretty weird because of the function signature. 

Anyways, I've refactored it a little bit, now bypassing __init__ via calling __new__ directly when I need to construct from data without triggering a reload (when slicing) and providing the user with a simple filename constructor.

Reply:
👍